### DISCUSSION TITLE
ZigBee Coordinator cod.m CC2652P7 with Tasmota ?

### DISCUSSION THREAD

**Discussion by Noschvie** [arendst/Tasmota#21969] [General]
Hello

found this [hardware](https://shop.codm.de/automation/zigbee/40/zigbee-coordinator-cc2652p7) designed by Patrik Mayer, @codmpm.

Features

- CC2652P7 Texas Instruments Multiprotokoll 2.4GHz Funkmodul
- ZigBee 3.x, Z-Stack Firmware (Koenkk) Kompatibel mit zigbee2mqtt (z2m), Home Assistant (zha), ioBroker, etc.
- Stromversorgung über PoE (802.3af) oder USB-C, <1W Stromverbrauch
- ESP32 Open Source Firmware (cod.m ZigStar UZG Fork)

Is Tasmota able to support this hardware and be uses as platform for this?
Thanks!

**Comment by Jason2866** [comment]
Since the ZbBridgePro uses the this zigbee chip and is supported from Tasmota the hardware you mentioned should work.

**Comment by s-hadinger** [comment]
Not cheap but I suppose Poe is the reason of such a price. 
indeed it should work nicely with Tasmota Z2T or Z2M/ZHA in tunnel mode

**Comment by codmpm** [comment]
Thank you @Noschvie! I'm willing to test this. 
I must admit, that did not tip my toes into Tasmota and ZigBee... only used it on shellies and self built esp hardware..

Can someone put me in the right direction?


> Not cheap but I suppose Poe is the reason of such a price.

Part of... and CE, WEEE, VERE, etc. 
And completely developed and made in Germany 🇩🇪😉



Btw, there's also an english product page and english documentation: 
https://shop.codm.de/en/automation/zigbee/40/zigbee-coordinator-cc2652p7-PoE
https://docs.codm.de/en/zigbee/coordinator/

**Reply by s-hadinger** [reply]
> Can someone put me in the right direction?

What do you mean?

**Reply by codmpm** [reply]
Like where to find FW to test, summary Page for what it's all about?
As I said, had nothing to do with Tasmota in combination with ZigBee, yet.

**Reply by Jason2866** [reply]
@codmpm May you start reading here. https://templates.blakadder.com/sonoff_ZBBridge-P.html  Flashing should at least. Probably the GPIOs for the connections of the zigbee chips needs adopting

**Reply by s-hadinger** [reply]
And of course https://tasmota.github.io/docs/Zigbee/

**Comment by codmpm** [comment]
Ok... the ZBPro seems to be ESP8266... ours is ESP32...

https://templates.blakadder.com/sonoff_ZBBridge.html - will try to write a template for the CZC.
What is I2C used for in the ZBPro?

**Reply by s-hadinger** [reply]
No zbpro is esp32

**Reply by codmpm** [reply]
Sorry.. wrong blakadder template.

**Comment by codmpm** [comment]
Can someone clarify what will be the benefits from using Tasmota over our own Firmware?

**Reply by s-hadinger** [reply]
If you don't want to spend years in coding. Sorry I don't understand : Tasmota compared to what?

**Reply by codmpm** [reply]
To our own already finished firmware which is running on the device and - to my understanding - does exactly the same: 
https://github.com/codm/czc-firmware/


It is currently based on the UZG firmware (https://github.com/mercenaruss/uzg-firmware) and we are working on a new version based on the XZG firmware (https://github.com/xyzroe/XZG).
Our CZC is already supported by the XZG firmware.

We already have autodiscovery for ZHA and mdns for z2m.
![codm-zigbee-coordinator-webif](https://github.com/user-attachments/assets/2cc442b0-0c0f-4312-975f-de5c867e0845)

**Reply by Jason2866** [reply]
@codmpm If you have Tasmota on all other devices, you will like the choice not to have to use a other software ;-)

EDIT: Fun fact the used Platformio Arduino framework of the projects you mentioned is from Tasmota.

**Reply by codmpm** [reply]
Ok. Whats about the all present warnings to not use proxied ZigBee over WiFi, VPN, etc? The ZBPro and default Tasmota use only WiFi. Albeit I will check to configure the LAN8720 with Tasmota.

See 
https://www.zigbee2mqtt.io/advanced/remote-adapter/connect_to_a_remote_adapter.html#connect-to-a-remote-adapter

>Be aware that it is not recommended to use a Zigbee Coordinator via a Serial-Proxy-Server (also known as Serial-to-IP bridge or Ser2Net remote adapter) over a WiFi, WAN, or VPN connection.
>Serial protocols used by Zigbee Coordinator do not have enough robustness, resilience, or fault-tolerance to handle packet loss and latency delays that can occur over unstable connections.
>Zigbee Coordinator requires a stable local connection to its serial port interface with no drops in communication between it and the Zigbee gateway application running on the host computer.
>Thus be warned that connecting to a network-attached remote Zigbee Coordinator over WiFi/WAN/VPN using Ser2Net or other Serial Proxy/Forwarding Tunnel is not supported for normal operation.

and 
https://www.home-assistant.io/integrations/zha#warning-about-using-zigbee-coordinator-over-wi-fiwanvpn

> Be aware that using a Zigbee Coordinator via a Serial-Proxy-Server (also known as Serial-to-IP bridge or Ser2Net remote adapter) over a Wi-Fi, WAN, or VPN connection is not recommended.
> Serial protocols used by the Zigbee Coordinator do not have enough robustness, resilience, or fault tolerance to handle packet loss and latency delays that can occur over unstable connections.
> A Zigbee Coordinator requires a stable local connection to its serial port interface with no drops in communication between it and the Zigbee gateway application running on the host computer.

**Reply by s-hadinger** [reply]
Ah ok. I didn't know your own firmware. Then it's a question of choice. 

It's definitely not ideal to have wifi and zigbee active at the same time but it works reasonably well. This is why I prefer ethernet based controllers

**Comment by codmpm** [comment]
Tried to get it working, wrote a template: 

```{"NAME":"CZC-1.0","GPIO":[1,1,1,1,3552,1,1,1,576,1,288,1,5792,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,5793,1,1,32,3584,0,0,1],"FLAG":0,"BASE":1}```

Result: 
```
10:53:55.953 RSL: STATE = {"Time":"2024-08-16T10:53:55","Uptime":"0T00:00:09","UptimeSec":9,"Heap":169,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":29,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":27},"Wifi":{"AP":1,"SSId":"codm-i","BSSId":"7E:A7:41:74:D4:FF","Channel":1,"Mode":"HT20","RSSI":66,"Signal":-67,"LinkCount":1,"Downtime":"0T00:00:04"}}
10:54:02.563 ZIG: rebooting ZNP device
10:54:02.763 ZbInput discarding byte 00
10:54:07.263 RSL: RESULT = {"ZbState":{"Status":1,"Message":"CCxxxx ZNP booted","RestartReason":"Power-up","MajorRel":2,"MinorRel":7}}
10:54:07.463 RSL: RESULT = {"ZbState":{"Status":50,"MajorRel":2,"MinorRel":7,"MaintRel":1,"Revision":20230507}}
10:54:07.613 RSL: RESULT = {"ZbState":{"Status":2,"Message":"Resetting configuration"}}
10:54:12.165 RSL: RESULT = {"ZbState":{"Status":3,"Message":"Configured, starting coordinator"}}
10:54:17.414 RSL: RESULT = {"ZbState":{"Status":40,"NewState":9,"Message":"Started as coordinator"}}
10:54:17.515 RSL: RESULT = {"ZbState":{"Status":51,"IEEEAddr":"0x00124B002A7B4A9C","ShortAddr":"0x0000","DeviceType":7,"DeviceState":9,"NumAssocDevices":0}}
10:54:18.014 RSL: RESULT = {"ZbState":{"Status":0,"Message":"Started"}}
```

Also put the rule ```backlog rule1 on system#boot do TCPStart 6608 endon ; rule1 1 ; TCPStart 6608``` in the console to start the TCP server. Sadly no luck.

```
10:56:13.821 CMD: backlog rule1 on system#boot do TCPStart 6608 endon ; rule1 1 ; TCPStart 6608
10:56:13.837 RUL: Stored uncompressed, would compress from 37 to 26 (-30%)
10:56:13.839 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":37,"Free":474,"Rules":"on system#boot do TCPStart 6608 endon"}}
10:56:14.069 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":37,"Free":474,"Rules":"on system#boot do TCPStart 6608 endon"}}
10:56:14.338 RSL: RESULT = {"Command":"Unknown"}
```

It seems to be very difficult to get my head around this, if you are not familar with the Tasmota universe. How do I enable the TCP port correctly?

![Bildschirmfoto 2024-08-16 um 11 54 31](https://github.com/user-attachments/assets/d2a91608-be40-4276-bb49-d01c4c8e2a47)

**Reply by s-hadinger** [reply]
You are mixing up modes. Either you have zigbee managed by Tasmota (aka Z2T) which is what you have in the screenshot. Or you just setup a tcp tunnel for Zigbee2Mqtt or ZHA. In such case the template is different: use TCP Rx and TCP Tx

**Reply by codmpm** [reply]
Sorry... thats very confusing. I followed Manual Confgiuration from here (https://zigbee.blakadder.com/Sonoff_ZBBridge-P.html) with a modified template to adapt the GPIO's of our board.

Did I flash the wrong firmware? I used "Tasmota Zigbee" from https://tasmota.github.io/install/

**Reply by sfromis** [reply]
With ESP32+CC2652, I'm using `tasmota32-zbbrdgpro.bin` which is selection "Sonoff-ZigbeeBridgePro" in the Tasmota Web Installer. This is also the one recommended on https://templates.blakadder.com/sonoff_ZBBridge-P.html

And when it comes to template and gpio config, the standard procedure is for "local" Zigbee2Tasmota, where Tasmota manages the coordinator, and "bridges" to/from MQTT payloads (or local coding). Then it is not relevant to mess with TCPStart etc, as it is just standard Tasmota networking.

If you **instead** do not want Tasmota to have anything to do with Zigbee (after flashing the Zigbee firmware), you can reconfigure to have ser2net (serial via TCP), allowing Zigbee2MQTT or Home Assistant ZHA, to take on the role of managing the coordinator function. In this case it is just a TCP tunnel to serial access to the CC2652, as mentioned by s-hadinger.

**Reply by codmpm** [reply]
Do I have to enable the TCP Port by hand or is it enabled by default? If yes, which port is it?

**Reply by s-hadinger** [reply]
The port is selected via the commands you used above "tcpstart"
The command failed becaused TCP GPIOs were not configured

**Reply by Jason2866** [reply]
There is a write up how to setup -> https://zigbee.blakadder.com/Sonoff_ZBBridge-P.html
GPIOs needs to be set in the Template reflecting your hardware!

**Reply by codmpm** [reply]
I'm aware of that @Jason2866, see my defined template in the post of this reply.

Did not know that I have to define TCP GPIO's, though. 
How do I configure these? Set them to an unused GPIO? As said, very confusing.
As we use the LAN8720 there are barely empty IO's on the  ESP anymore.

**Reply by sfromis** [reply]
You do not have to use TCP, unless you do not want to use Tasmota for Zigbee, outside of providing the tunnel. But if this is what you want, you should reconfigure the Zigbee Tx/Rx pins to instead be TCP.

**Reply by codmpm** [reply]
Sorry.. I don't quite understand.
What is "Tasmota for ZigBee"? Thought there is just the tunnel part?

**Reply by s-hadinger** [reply]
No. There is full support for zigbee locally from Tasmota, including mqtt but also bridging to Alexa Hue emulation. And very soon Matter - Zigbee support. 

Due to the limitations of esp8266 and esp32, it is obviously not as rich as Zigbee2Mqtt, but it doesn't require you to manage a separate system

**Reply by s-hadinger** [reply]
Now I understand better your question about firmwares. If all you need is tcp to serial bridge, it's much easier

**Reply by codmpm** [reply]
To be honest, I don't need Tasmota. Our FW does already do tcp-to-serial with a nice GUI, etc ;-) We are working on a Thread FW for the CC chip.
Just wanted to check and try if the Tasmota savy people can flash it to our hardware and use it.

I don't know which is better. tcp-to-serial or full local zigbee support in the Tasmota world.
Happy to show the schematics. It's an ESP32 with LAN8720, FT231X USB-Serial and CC2652P7 in a - to me - nice form factor.
Is this usabel for anything Tasmota related?

**Reply by s-hadinger** [reply]
I suppose yes that it could interest the Tasmota community. The closest currently is Athom from China but without POE. 

Would you be kind enough to give away one device so I  can create a 1-click autoconfiguration file?

**Reply by sfromis** [reply]
When you got to the point of Tasmota reporting the progress messages about the CC2652 booting, it does look like the device is workable with Tasmota too. Arguably, it is not much in "verification", but as the P7 variant should not be much of a difference, I'd expect it to work as any other CC2652. Not sure how much of a benefit the extra RAM would be in this scenario, if any.

**Reply by s-hadinger** [reply]
If there is extra ram, it should be usable by Koenkk firmware to support more devices

**Reply by codmpm** [reply]
> Would you be kind enough to give away one device so I can create a 1-click autoconfiguration file?

Of course. Hit me up at via email.
There are many alternatives out there (UZG, SMLight Products, TubesZB, etc.) but they typically are not open source hardware and have to be reverse engineered. I can give you the schematics (CC-BY-NC-SA) of our CZC.

@sfromis We built a CC2652P variante before... and the first prototype was with the P instead of the P7. They are interchangeable from the point of view of the serial connection.
Here's an overview of the CC chips: https://www.ti.com/technologies/wired-wireless-connectivity/matter/overview.html


Has Tasmota any way to updaten the firmware of the CC chip? 
In the next CZC Firmware (based on XZG) this could be done from the webinterface.

**Reply by s-hadinger** [reply]
Sure will do. 

Yes Tasmota has support for BSL mode and flashing hex files. It's via Berry scripts and not a simple click on the web UI. But until now it was rare to upgrade the zigbee firmware

**Comment by Noschvie** [comment]
Thank you for this conversation and discussion that brings together great software with great hardware!

**Comment by s-hadinger** [comment]
I received my device, and already fixed a problem. The CC2652P7 takes longer than usual so I increased the boot timeout from 5s to 10s.

Unfortunately I can't find the schematics of the device to configure Ethernet.

**Comment by s-hadinger** [comment]
I finally found the GPIOs in the firmware source.

There is now a new autoconf file:
- configure ZIgbee and Ethernet
- add Ethernet Network Flipper to disconnect Wifi when Eth is on, and avoid interference with Wifi

<img width="353" alt="image" src="https://github.com/user-attachments/assets/9860bcf7-3153-4764-a780-e11f2d672e4e">

**Comment by s-hadinger** [comment]
I also added the following for ZHA / Zigbee2MQTT (port 8888):

<img width="353" alt="image" src="https://github.com/user-attachments/assets/8ead4ec4-269b-4597-9842-c5818f5aaa58">

**Comment by codmpm** [comment]
Thank you very very much!
